### ISSUE TITLE
Is nvidia container runtime necessary for app pod when using cdi-annotations strategy?

### ISSUE THREAD

**Issue by xiongzubiao** [NVIDIA/k8s-device-plugin#481]
The plugin is set to use `cdi-annotations` strategy. The `nvidia` container runtime is added to CRI-O but not set as default. When I deploy a pod requesting gpu, I can see that some cdi annotations are added to the container spec (config.json):
```json
                "io.kubernetes.cri-o.Annotations": "{\"io.kubernetes.container.hash\":\"c89ea172\",\"io.kubernetes.container.restartCount\":\"0\",\"io.kubernetes.container.terminationMessagePath\":\"/dev/termination-log\",\"io.kubernetes.container.terminationMessagePolicy\":\"File\",\"io.kubernetes.pod.terminationGracePeriod\":\"30\",\"nvidia.cdi.k8s.io/nvidia-device-plugin_9f431d3c-53a7-4e3c-b4dd-8283d819d109\":\"k8s.device-plugin.nvidia.com/gpu=GPU-655005fd-ef3b-d676-4606-ba4bb3f75ba5\"}",
                "nvidia.cdi.k8s.io/nvidia-device-plugin_9f431d3c-53a7-4e3c-b4dd-8283d819d109": "k8s.device-plugin.nvidia.com/gpu=GPU-655005fd-ef3b-d676-4606-ba4bb3f75ba5",
```
 , but the those mounts from the cdi spec are not added.

It looks like I still need to use the `nvidia` container runtime (mode = "cdi") to make the mounts show up in the container's spec. Isn't CRI-O supposed to do that, as it knows where to find the cdi spec?

**Comment by xiongzubiao** [comment]
Okay, this time I tried to deploy the plugin without using nvidia gpu operator, by following https://gitlab.com/nvidia/kubernetes/device-plugin/-/merge_requests/298. It works fine. All mounts are added correctly even not using nvidia runtimeclass.

The annotations in the container spec are:
```json
                "io.kubernetes.cri-o.Annotations": "{\"cdi.k8s.io/nvidia-device-plugin_b6fc205c-36a6-4096-8533-13b3b94b1dd8\":\"k8s.device-plugin.nvidia.com/gpu=GPU-655005fd-ef3b-d676-4606-ba4bb3f75ba5\",\"io.kubernetes.container.hash\":\"e20e5814\",\"io.kubernetes.container.restartCount\":\"0\",\"io.kubernetes.container.terminationMessagePath\":\"/dev/termination-log\",\"io.kubernetes.container.terminationMessagePolicy\":\"File\",\"io.kubernetes.pod.terminationGracePeriod\":\"30\"}",
                "cdi.k8s.io/nvidia-device-plugin_b6fc205c-36a6-4096-8533-13b3b94b1dd8": "k8s.device-plugin.nvidia.com/gpu=GPU-655005fd-ef3b-d676-4606-ba4bb3f75ba5",
```
Looks like the main difference is the prefix: `nvidia.cdi.k8s.io` vs. `cdi.k8s.io`.

Why does the gpu operator use a different prefix? If that is the reason, it sound like a bug of the gpu operator. BTW, why does the plugin even allow to change the annotation prefix?

**Comment by xiongzubiao** [comment]
Hi @klueska, should I open an issue in Nvidia's GPU Operator repo, since it sounds like a problem of the operator?

**Comment by elezar** [comment]
@xiongzubiao it is not required to use the NVIDIA Container Runtime when using native CDI support in a container engine such as cri-o or containerd. As you have already demonstrated when not using the GPU Operator.

The reason for using the `nvidia.cdi.k8s.io` prefix over the `cdi.k8s.io` that is enabled by default by CRI-O is that we are required to support a wide range of container engine versions. To do so we use the NVIDIA Container Runtime (in CDI mode) to make the required modifications to the incoming OCI runtime specification. The custom prefix ensures that these modifications are not ALSO made by Containerd or Crio.

We are looking at the implications of making use of native CDI support in engines such as containerd or CRI-O instead.

**Comment by xiongzubiao** [comment]
@elezar That makes sense. Thanks for the explanation.

It will be great if the difference of the annotation prefix could be documented somewhere. One might simply think that the NVIDIA runtime is not needed as long as the CRI runtime supports CDI, just like me :-).

**Comment by elezar** [comment]
Yes, we have a documentation MR out [here](https://gitlab.com/nvidia/kubernetes/device-plugin/-/merge_requests/298) that we need to review and integrate as we continue testing native CDI support in various runtimes.
